src/coreclr/jit/llvm.cpp

```diff
@@ -1393,6 +1406,74 @@ void Llvm::importStoreInd(GenTreeStoreInd* storeIndOp)
     }
 }
 
+void Llvm::storeObjAtAddress(Value* baseAddress, Value* data, unsigned startIx, unsigned gcFieldCount, FieldStoreLayout* gcFieldLayout)
+{
+    unsigned i = startIx;
+    while (!gcFieldLayout[i].IsEndStruct)
+    {
+        Value* address = _builder.CreateGEP(baseAddress, _builder.getInt32(gcFieldLayout[i].FieldOffset));
+
+        // get the data if this instruction has data
+        Value* fieldData = nullptr;
+        if (!gcFieldLayout[i].IsEndStruct)
+        {
+            if (data->getType()->isStructTy())
+            {
+                unsigned index = _module->getDataLayout().getStructLayout((llvm::StructType*)(data->getType()))->getElementContainingOffset(gcFieldLayout[i].FieldOffset);
+                fieldData = _builder.CreateExtractValue(data, index);
+            }
+            else
+            {
+                // single field IL structs are not LLVM structs
+                fieldData = data;
+            }
+        }
+
+        if (gcFieldLayout[i].IsStartStruct)
+        {
+            storeObjAtAddress(address, fieldData, i + 1, gcFieldCount, gcFieldLayout);
+        }
+        else
+        {
+            _builder.CreateCall(getOrCreateRhpAssignRef(),
+                                ArrayRef<Value*>{address, castIfNecessary(fieldData, Type::getInt8PtrTy(_llvmContext))});
+        }
+        i++;
+    }
+}
+
+void Llvm::importStoreObj(GenTreeStoreInd* storeIndOp)
+{
+    unsigned gcFieldCount;
+    Value* baseAddress = getGenTreeValue(storeIndOp->Addr());
+
+    if (!storeIndOp->OperIsBlk())
+    {
+        failFunctionCompilation(); // cant get struct handle. TODO-LLVM: try as an assert
+    }
+
+    ClassLayout* structLayout = storeIndOp->AsBlk()->GetLayout();
+
+    // zero initialization  check
+    GenTree* dataOp = storeIndOp->Data();
+    if (dataOp->OperIs(GT_CNS_INT) && dataOp->AsIntCon()->IconValue() == 0)
+    {
+        _builder.CreateMemSet(baseAddress, _builder.getInt8(0), _builder.getInt32(structLayout->GetSize()), {});
+        return;
+    }
+
+    CORINFO_CLASS_HANDLE structClsHnd = structLayout->GetClassHandle();
+    FieldStoreLayout* gcFieldLayout;
+    gcFieldCount = _getObjectLayout(_thisPtr, structClsHnd, &gcFieldLayout);
```

Comment:
Re-creating this for each node seems expensive. We need a side table that'll associate it with `ClassLayout*` pointers.